docs/design/2024-09-06-maintain-priority-queue-in-memory.md

```diff
@@ -0,0 +1,567 @@
+Right now, we only have an in-memory channel to send the DDL events to the statistics module from the DDL module. So we need to find a reliable way to store and fetch DDL events.
+
+To ensure that DDL events are processed correctly, DDL module writes them to a physical table with preset primary key in the same transaction of finishing DDL. The stats module reads these events sorted by their primary key, leveraging the fact that DDL events are executed in order, which guarantees that their IDs are also ordered.
+We use a large transaction to process the DDL events within the stats module and delete them once processed. This approach eliminates the need for specific checkpoints during failover and guarantees exactly once delivery, as each event is read and processed only once.
+
+### The New Priority Queue - Maintain the heap in memory
+
+Because we are constantly adding, removing, and updating the heap, we need to use a thread-safe heap to store analysis jobs.
+The required interfaces look like the following:
+
+| Operation | Time complexity |
+| --------- | --------------- |
+| Add       | O(log n)        |
+| Update    | O(log n)        |
+| Delete    | O(log n)        |
+| Pop       | O(1)            |
+| Get       | O(1)            |
+
+#### Information to be stored
+
+```go
+type Indicators struct {
+        // ChangePercentage is the percentage of the changed rows.
+        // Usually, the more the changed rows, the higher the priority.
+        // It is calculated by modifiedCount / last time analysis count.
+        ChangePercentage float64
+        // TableSize is the table size in rows * len(columns).
+        TableSize float64
+        // LastAnalysisDuration is the duration from the last analysis to now.
+        LastAnalysisDuration time.Duration
+}
+
+type AnalysisJob interface {
+        // SetWeight sets the weight of the job.
+        SetWeight(weight float64)
+
+        // GetWeight gets the weight of the job.
+        GetWeight() float64
+
+        // HasNewlyAddedIndex checks whether the job has a newly added index.
+        HasNewlyAddedIndex() bool
+
+        // GetIndicators gets the indicators of the job.
+        GetIndicators() Indicators
+}
+```
+
+#### Basic API
+
+```go
+type Heap[T any, K comparable] interface {
+    // Add/Update elements
+    Add(obj T) error
+    BulkAdd(list []T) error
+    AddIfNotPresent(obj T) error
+    Update(obj T) error // Same as Add
+
+    // Remove element
+    Delete(obj T) error
+
+    // View/Pop top element
+    Peek() (T, error)
+    Pop() (T, error)
+
+    // Query operations
+    List() []T
+    ListKeys() []K
+    Get(obj T) (T, bool, error)
+    GetByKey(key K) (T, bool, error)
+
+    // State operations
+    Close()
+    IsClosed() bool
+}
+
+// Constructor
+func NewHeap[T any, K comparable](keyFunc KeyFunc[T, K], lessFunc LessFunc[T]) *Heap[T, K]
+
+// Helper function type definitions
+type KeyFunc[T any, K comparable] func(T) (K, error)
```

Comment:
Why it will return error? I think ID of `T` is preset in `T` and will not have error when reading it.

Reply:
Just for the sake of scalability, it's not always a simple number.